### PR TITLE
gnrc_netif: make use of counting event

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

> Can't we have a counting event type?

This adds a counting event type. If multiple events of the same type occur, the callback function is called for each of them instead of being called only once.

This solves the issue where if a network driver has a queue and generates an event for each queued packet, those packets will be stuck in the queue if gnrc_netif is not fast enough to process them in time before the next event occurs.

### Issues/PRs references

alternative to #18204, #18201

### CODE REVIEW

File: sys/event/event.c [hunk 1]
```diff
@@ -60,29 +88,79 @@ void event_cancel(event_queue_t *queue, event_t *event)
...
-event_t *event_wait_multi(event_queue_t *queues, size_t n_queues)
+static event_counted_t *_get_counted(event_queue_t *queue)
+{
+    clist_node_t *node = clist_lpeek(&queue->event_list);
+    event_counted_t *result = container_of(node, event_counted_t, list_node);
+
+    if (result) {
+        assert(result->num_posted);
+        if (--result->num_posted == 0) {
+            clist_lpop(&queue->event_list);
```
Comment (maribu) [inline sys/event/event.c:127]:
```suggestion
            clist_lpop(&queue->event_list);
            result->list_node.next = NULL;
```

**Review by maribu — COMMENTED** [review]
Code looks good to me.

I would however prefer if this would introduce an `event_queue_counted_t`. Ideally such that implicit casts from `event_queue_t *` to `event_queue_counted_t *` and vice versa are impossible. This would make it easy to detect when one switches from one event type to the other and forgets to update a function call.

**Comment by maribu** [comment]
I changed my mind. Considering that we spend quite some effort on having the event handler thread(s) being merged to safe stack space, it is counter intuitive to now double the number of worker threads needed to handle both counting and non-counting events.

Could we instead just enqueue both counting and non-counting events into the same queue and use [pointer tagging](https://doc.riot-os.org/group__sys__ptrtag.html) to mark those which are counting? In case counting events are not used (the module is not included), no pointer tagging is needed and no overhead is introduced. By using pointer tagger this would not blow up the size of `event_t`.

File: sys/event/event.c (hunk 1 above)
Comment (benpicco) [reply sys/event/event.c:127]: Not really necessary, but I suppose it doesn't hurt

Comment (maribu) [reply sys/event/event.c:127]: This is done for the non-counting event. IMO, we should have this consistent. (And maybe there is code that checks if an event is queued by comparing the `next` pointer with `NULL`.)

**Comment by jia200x** [comment]
I definitely like the idea of having a "counter" event. However, after the [discussed use case](https://github.com/RIOT-OS/RIOT/pull/18201) in `gnrc_netif`, I think we cannot guarantee a deterministic behavior for all devices when there are multiple RX Done events without some cooperation from the device drivers.

As it is now, the `NETDEV_EVENT_ISR` event is used to offload the `_isr` function to thread context. Therefore, this follows a semantic of "please offload asap" more than "process interrupts in the order they arrive". as the information for the latter case is discarded when calling `_isr`. Therefore, _if_ a device is able to queue several RX frames before ISR offloading, the device will need to keep state of this anyway.

Therefore, I would be in favour of:
1. Merging this PR anyway, as it enables some use cool cases
2. Simply take https://github.com/RIOT-OS/RIOT/pull/18201 for the multiple RX frames before ISR offloading.

Opinions?

**Comment by benpicco** [comment]
#18201 would certainly be the least invasive change. My concern for creating this PR was that I imagined many other places, where event users might be surprised by the singleton nature of posting events and having a solution ready for those subtle bugs.

For SLIP (and ethos) it would certainly be an argument to move forward with #18066 then as those would then also benefit from the common infrastructure in the chunked ringbuffer for queued frames.

**Comment by jia200x** [comment]
>  My concern for creating this PR was that I imagined many other places, where event users might be surprised by the singleton nature of posting events and having a solution ready for those subtle bugs

Yup. As mentioned early, I still think this is a good idea as a module. We should definitely indicate the singleton behavior of normal events though.

**Comment by jia200x** [comment]
> Merging this PR anyway, as it enables some use cool cases

Sorry, I confused this PR. I meant, merging this mechanism (#18204 ), not the GNRC netif mechanism

**Comment by benpicco** [comment]
Now I'm confused. #18204 makes all events counting and thus changes the API which may have undesired consequences.

Instead this PR introduces a separate counting event type (and tries to share most code with normal event) and makes use of that in gnrc_netif.

**Comment by jia200x** [comment]
Ah sorry, you are right. What I meant is merging the counting mechanism, but not the usage on GNRC, as it's mainly used for IRQ offloading.

**Comment by benpicco** [comment]
Now that it's the responsibility of the driver to handle this, we can close this PR.

**Pull request closed without merging**